admin/server/device_code.go

```diff
@@ -0,0 +1,315 @@
+}
+
+// handleDeviceCodeRequest creates a 24 digit random device code and 8 digit user code and returns that
+// to the client. The device code is used to poll for an access token, and the user code is displayed
+// to the user and is used to authorize the device. The device code and user code are stored in the
+// server's device code store.
+func (s *Server) handleDeviceCodeRequest(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		s.logger.Error("failed to read request body", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	bodyStr := string(body)
+	values, err := url.ParseQuery(bodyStr)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	clientID := values.Get("client_id")
+	if clientID == "" {
+		http.Error(w, "client_id is required", http.StatusBadRequest)
+		return
+	}
+	scopes := strings.Split(values.Get("scope"), " ")
+	if len(scopes) == 0 {
+		http.Error(w, "scope is required", http.StatusBadRequest)
+		return
+	}
+	if len(scopes) > 1 || scopes[0] != "full_account" {
+		http.Error(w, "invalid scope", http.StatusBadRequest)
+		return
+	}
+	authCode, err := generateDeviceAndUserCode()
+	if err != nil {
+		s.logger.Error("failed to generate device and user code", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	authCode.ClientID = clientID
+	err = s.admin.DB.InsertAuthCode(r.Context(), authCode)
+	if err != nil {
+		s.logger.Error("failed to insert auth code", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	verificationURI := s.conf.DeviceVerificationHost + "/oauth/device"
+	resp := DeviceCodeResponse{
+		DeviceCode:              authCode.DeviceCode,
+		UserCode:                authCode.UserCode,
+		VerificationURI:         verificationURI,
+		VerificationCompleteURI: verificationURI + "?user_code=" + authCode.UserCode,
+		ExpiresIn:               tokenExpirationSeconds,
+		PollingInterval:         5,
+	}
+	e := json.NewEncoder(w)
+	e.SetIndent("", "  ")
+	err = e.Encode(resp)
+	if err != nil {
+		s.logger.Error("failed to encode response", zap.Error(err))
+		internalServerError(w, err)
+	}
+}
+
+// handleUserCodeConfirmation handles the user code confirmation page. The user code is displayed
+// to the user and they are asked to confirm that they want to authorize the device. If the user
+// confirms, the device code is marked as approved in the server's device code store.
+func (s *Server) handleUserCodeConfirmation(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	userCode := r.URL.Query().Get("user_code")
+	if userCode == "" {
+		http.Error(w, "user_code is required", http.StatusBadRequest)
+		return
+	}
+	confirmation := r.URL.Query().Get("code_confirmed")
+	if confirmation == "" {
+		http.Error(w, "no code confirmation", http.StatusBadRequest)
+		return
+	}
+	sess, err := sessions.NewCookieStore([]byte(s.conf.SessionSecret)).Get(r, authSessionName)
```

Comment:
@begelundmuller how do I access constant https://github.com/rilldata/rill-developer/blob/main/admin/server/auth/handlers.go#L17 here ? I tried even exporting by capitalizing `C` but did not work

Reply:
Capitalizing it should work, called as `auth.CookieName` – maybe you're important the wrong `auth` package?

However, I feel the auth cookie should not be accessed outside that package. Probably the device flow code should be moved to inside the `admin/server/auth` package instead? And then add the endpoints in the `RegisterEndpoints` function.